VoxelHashMap ereasing of the voxels

kulfer80 opened the pull request:
## This is a proposed solution that fixes #309 
#### The reason for this pull request is to consider the safety of the code so that it is not susceptible to potential unexpected behavior.

kulfer80 commented:
@nachovizzo I did more testing and investigations, this is a disassembled code:
https://godbolt.org/z/GrhMve1eh

---

In this scenario it can be clearly seen that by original solution the one key,value pair isn't deleted that should have been deleted by condition.

kulfer80 commented:
@nachovizzo I have just recently tested both implementations on KITTI_raw dataset with robin_map and in some cases voxels were skipped and they weren't removed, in most of the cases size of map remained same in both implementations. 
@nachovizzo Should I send you some logs or testing scenarios ? Let me know.

kulfer80 commented:
> ```c++
> for (auto it = map_.begin(); it != map_.end();) {
>     if ((pt - origin).squaredNorm() > (max_distance2)) {
>         map_.erase(it);
>     } 
>     ++it;
> }

In this version, what is happening is:
When an element is erased, the elements after it might get shifted, but the iterator keeps moving forward, this can lead to either skipped element or revisiting already processed element.

nachovizzo commented:
@kulfer80 yes and no :)

Although over the surface looks like what's happening, and indeed makes more sense after reading the documentation I still believe this is not the case. What I think it's true is that after `map_.erase(it)` that particular iterator is no longer valid, `it` points to an element that no longer exists. But since we do not use it again, it does not  change the results (to my eyes). After the `++it` then the iterator will point to the next valid element in the hash table and thus, valid again. 

I would agree that this is kind of "hacky" since we rely on not using such iterator in between the loop, although with the C++17 structured bindings used in the original implementation that iterator is only visible to the compiler and not to the user. So it's should be safe.

I did this very dirty program:
```cpp
#include <iostream>
#include <tsl/robin_map.h>

int main() {
  tsl::robin_map<int, int> map = {
      {1, 10},
      {2, 20},
      {3, 30},
  };
  std::cout << "-------------------------------------" << std::endl;
  for (auto it = map.begin(); it != map.end();) {
    const auto &[key, value] = *it;

    if (value > 10) {
      std::cout << "value before " << std::get<1>(*it) << std::endl;
      std::cout << "size  before " << map.size() << std::endl;
      map.erase(it);
      std::cout << "value  after " << std::get<1>(*it) << std::endl;
      std::cout << "size   after " << map.size() << std::endl;
    }
    ++it; // here the iterator gets updated and no longer invalid
    if (value > 10) {
      std::cout << "value    end " << std::get<1>(*it) << std::endl;
      std::cout << "size     end " << map.size() << std::endl;
      std::cout << "-------------------------------------" << std::endl;
    }
  }
  return 0;
}
```

And this is the result, that somehow supports my shady theory. After erasing the iterator is invalid, but after incrementing it everything goes back to normal

![image](https://github.com/PRBonn/kiss-icp/assets/21349875/dde7d480-3211-4a96-965d-148c11a5bb65)

Do you have any solid tests that confirm this is a real problem?

Thanks for the help bwt!

kulfer80 commented:
@nachovizzo You are right about validating iterator. The thing is, the original solution that is now implemented in the code would work if every voxel had unique hash, but it some cases it isn't true. 

---

Imagine this situation you have robin_map that contains key, value pairs, in which key == value so simplicity.
The hash is key % 2, so there will be collisions. Now let s say condition is value > 5, we are currently at element with value 6, iterator is pointing at bucket let s say 0x555555573074, the thing that happens after erasing is the element with value 8 is moved to this bucket and we moved iterator to bucket 0x555555573080 so basically we skipped bucket that is occupied by element with value 8 and we finish iterations without removing it, even though condition says it should be removed.

The proposed solution handles it because it returns the same bucket if it is valid iterator if not it will advance to next element.

Here is the snippet that you can try on your own (I know it s ugly written):

```#include <iostream>
#include <tsl/robin_map.h>

struct MyHash {
    std::size_t operator()(int key) const {
        return key % 2;
    }
};

std::pair<int, int> original(int num) {
    tsl::robin_map<int, int, MyHash> map_;
    for (int i = 0; i < 13; ++i) {
        map_[i] = i;
    }

    int count = 0;
    for (auto it = map_.begin(); it != map_.end();) {
        const auto& [key, value] = *it;
        if (value > num) {
            map_.erase(it);
        }
        ++it;
        ++count;
    }

    return std::make_pair(map_.size(), count);
}

std::pair<int, int> proposed(int num) {
    tsl::robin_map<int, int, MyHash> map_;
    for (int i = 0; i < 13; ++i) {
        map_[i] = i;
    }

    int count = 0;
    for (auto it = map_.begin(); it != map_.end();) {
        const auto& [key, value] = *it;
        if (value > num) {
            it = map_.erase(it);
        } else {
            ++it;
        }
        ++count;
    }

    return std::make_pair(map_.size(), count);
}

int main() {
    const auto& [orig_size, orig_count] = original(5);
    const auto& [prop_size, prop_count] = proposed(5);
    std::cout << "Original size: " << orig_size << " count: " << orig_count << std::endl;
    std::cout << "Proposed: size: " << prop_size << " count: " << prop_count<< std::endl;
    return 0;
}

nachovizzo commented:
@kulfer80 :clap: 
Thanks for keep pushing!

I think now you are right :) So I'm happy to say I'd like to include this.

I don't think we are hitting this problem so many times, and I also don't think this will make any real impact on the system, since we are always working at the boundaries of the voxel map, and I never saw a "non removed" voxel block left behind in visualization

Nevertheless, I do value correct code. So let's move forward with it

The only thing I want to ask is to remove the `max_distance_2`. Just because it's unrelated to this change, after that, ready to go!

nachovizzo commented on cpp/kiss_icp/core/VoxelHashMap.cpp:
```diff
@@ -85,11 +85,13 @@ void VoxelHashMap::AddPoints(const std::vector<Eigen::Vector3d> &points) {
-        const auto max_distance2 = max_distance_ * max_distance_;
-        if ((pt - origin).squaredNorm() > (max_distance2)) {
-            map_.erase(voxel);
+        if ((pt - origin).squaredNorm() > (max_distance2_)) {
```
let's also put a `>=` just to avoid that small corner case where it's for whatever reason exactly the same as `max_distance` and we don't remove the voxel nor forward the iterator

nachovizzo commented:
> @kulfer80 👏 Thanks for keep pushing!
> 
> I think now you are right :) So I'm happy to say I'd like to include this.
> 
> I don't think we are hitting this problem so many times, and I also don't think this will make any real impact on the system, since we are always working at the boundaries of the voxel map, and I never saw a "non removed" voxel block left behind in visualization
> 
> Nevertheless, I do value correct code. So let's move forward with it
> 
> The only thing I want to ask is to remove the `max_distance_2`. Just because it's unrelated to this change, after that, ready to go!

@kulfer80 I did it myself, sorry :roll_eyes: 

Now waiting for the other maintainers review.

@tizianoGuadagnino , @benemer this PR is ready to review. Please take a deep look and give us your opinion :)

kulfer80 commented:
@nachovizzo It really occurred seldom, but if in future you change your mind and replace container robin_map with other it could lead to unexpected behavior. I only created max_distance2_ to not reintizialinzing this variable in every loop cycle. But you are right that way as you rewrote it, it is more readable. And thank you that we could have elaborated about this problem. One more time big thanks to your whole team that did outstanding work 👏.

nachovizzo commented:
> @nachovizzo It really occurred seldom, but if in future you change your mind and replace container robin_map with other it could lead to unexpected behavior. I only created max_distance2_ to not reintizialinzing this variable in every loop cycle. But you are right that way as you rewrote it, it is more readable. And thank you that we could have elaborated about this problem. One more time big thanks to your whole team that did outstanding work 👏.

feel free to open a 2nd PR about the max_distance. We could check, but I'd assume the compiler is optimizing that bit anyways

kulfer80 commented:
> > @nachovizzo It really occurred seldom, but if in future you change your mind and replace container robin_map with other it could lead to unexpected behavior. I only created max_distance2_ to not reintizialinzing this variable in every loop cycle. But you are right that way as you rewrote it, it is more readable. And thank you that we could have elaborated about this problem. One more time big thanks to your whole team that did outstanding work 👏.
> 
> feel free to open a 2nd PR about the max_distance. We could check, but I'd assume the compiler is optimizing that bit anyways

You are right I overthought it, I disassembled code and it gave the same result in both situations.

nachovizzo commented:
@kulfer80 congrats on your first contribution :partying_face: !!
